lib/web_apis/web_apis.es6.js

```diff
@@ -0,0 +1,347 @@
+          browserName: browserArr[0],
+          browserVersion: browserArr[1],
+          OS: browserArr[2],
+          OSVersion: browserArr[3],
+        }).id;
+      },
+    },
+    {
+      name: 'toMap_',
+      documentation: `Takes an array of browserAPIs and an array of
+        browser keys, produce a JSON containing browser and interface
+        relationship`,
+      args: [
+        {
+          name: 'browserAPIs',
+          typeName: 'BrowserAPI Array',
+          documentation: `An array returned from DAO.select.`,
+        },
+        {
+          name: 'brwoserKeys',
+          typeName: 'StringArray',
+          documentation: `A list of browser keys that contained in the
+            browserAPIs`,
+        },
+      ],
+      returns: {
+        typeName: 'JSON',
+        documentation: `Returns a JSON of the form:
+              {interfaceName: {APIName: [Boolean ...]} ...}
+            It contains information about browsers and their interface/API.`,
+      },
+      code: function(browserAPIs, browserKeys) {
+        return new Promise(function(resolve) {
+          let apiMap = {};
+          let browserDict = {}; // O(1) lookup for browser's index in header.
+          apiMap._header = browserKeys.slice();
+          let numBrowsers = apiMap._header.length;
+          // Fill in browserDict.
+          for (let i = 0; i < numBrowsers; i += 1) {
+            browserDict[this.getBrowserId_(browserKeys[i])] = i;
+          }
+          // Fill in apiMap.
+          let promises = [];
+          for (let i = 0; i < browserAPIs.length; i += 1) {
+            let interfaceId = browserAPIs[i].interfaceId;
+            promises.push(this.interfaces
+              .where(this.mlang.EQ(this.WebInterface.ID, interfaceId)).select()
+              .then((arrayDao) => {
+                let interfaceName = arrayDao.a[0].interfaceName;
+                let APIName = arrayDao.a[0].APIName;
+                if (!apiMap[interfaceName] ||
+                  typeof apiMap[interfaceName] !== 'object') {
+                  // If typeof is not object, this may be a build-in function.
+                  apiMap[interfaceName] = {};
+                }
+                if (!apiMap[interfaceName][APIName] ||
+                  typeof apiMap[interfaceName][APIName] !== 'object') {
+                  // If visit APIName for first time,
+                  // initialize an array of false.
+                  apiMap[interfaceName][APIName] = [];
+                  for (let k = 0; k < numBrowsers; k += 1) {
+                    apiMap[interfaceName][APIName].push(false);
+                  }
+                }
+                apiMap[interfaceName][APIName][
+                  browserDict[browserAPIs[i].browserId]] = true;
+              }));
+          }
+          Promise.all(promises).then(function() {
+            resolve(apiMap);
+          });
+        }.bind(this));
+      },
+    },
+    {
+      name: 'toMap',
```

Comment:
Let's discuss this and its helper offline. I don't think we've been over this design detail.